dask/tests/test_order.py

```diff
@@ -1221,7 +1219,8 @@ def test_array_vs_dataframe(optimize):
     diag_df = diagnostics(
         collections_to_dsk([mean.to_dask_dataframe()], optimize_graph=optimize)
     )
-    assert max(diag_df[1]) == max(diag_array[1])
```

Comment:
I think this is something we should look into. At least out of the box I do not understand the difference.

Arrays are doing worse here because they have a non-trivial tree reduction with some overlap

![image](https://github.com/user-attachments/assets/15bce82f-44d8-4c3d-a8b7-73d4da85c2ce)

while the dataframe (likely due to having a proper optimizer) are running a very ordinary tree reduction here 
![image](https://github.com/user-attachments/assets/067a0644-2543-4592-aaed-48ab9ff55264)


this feels off. @phofl any ideas why that is? (not a blocker, maybe a follow up)

Reply:
disabling optimization (entirely) on the array branch makes this a little easier to read

![image](https://github.com/user-attachments/assets/e2db36d8-f2f6-4624-b0d7-73df50318674)

Reply:
I guess the dataframe is somehow merging 
![image](https://github.com/user-attachments/assets/0f91361f-8d08-44d1-8543-dbe75223c4ec)

into a single task. That's still surprising me

Reply:
Maybe not running through his optimize call in the order test?

Reply:
I dug into this a bit and the way xarray does the conversion is that they put every variable into a dask Series. dask-expr calls optimize on every series instead of the concatenated result. This means that we load root nodes multiple times in some cases, creating a simpler topology and allowing for reduced pressure

Reply:
I wonder if we should change xarray do to the same... just call optimize on every variable instead of optimizing the entire thing

Reply:
That's a trade-off between repeatedly loading root nodes and having more memory consumption, ideally, we will be able to make this decision in array-expr at some point if we get the full thing

Reply:
Seems sketchy that two `random` calls get optimized to the same thing.

Here's a more realistic version with `zarr`. 

<details>
```python
import xarray as xr
import dask.array as da

size = 100
ds = xr.Dataset(
    dict(
        anom_u=(
            ["time", "face", "j", "i"],
            da.random.random((size, 1, 987, 1920), chunks=(10, 1, -1, -1)),
        ),
        anom_v=(
            ["time", "face", "j", "i"],
            da.random.random((size, 1, 987, 1920), chunks=(10, 1, -1, -1)),
        ),
    )
)
ds.to_zarr('temp.zarr', mode="w")
with xr.open_dataset('temp.zarr', chunks={}, inline_array=True) as ds:
    quad = ds**2
    quad["uv"] = ds.anom_u * ds.anom_v
    mean = quad.mean("time")
dask.visualize(mean, optimize_graph=True)
```

</details>

### `inline_array=True` 

With `inline_array=True`, the optimized graph looks like the one in this thread. 

![image](https://github.com/user-attachments/assets/0e376e5f-567d-4b94-b5d4-8fbd911c87a8)


### `inline_array=False`

We get something totally different!

![image](https://github.com/user-attachments/assets/77d39b60-17ce-4f61-87ab-f9735ee67aba)

Reply:
The topology looks very similar, the main difference seems to be that these data nodes at the bottom are cut off. We are not directly checking that they optimize to the same thing, we are checking if the topology is the same and if the ordering algorithm walks the graph in the same way.

We are using a think we call pressure here, that signals how many intermediate results you have to hold in memory to be able to process the whole graph.

The type of node that produces the input results (i.e. from zarr or random does not influence the result if the chunking is the same)

Reply:
Ok got it. Thanks for clarifying. No it does not matter.

We actually do the same thing as inline_array=True is doing for ordering, we cut these array nodes of before running through ordering to simplify the topology in these cases. So both cases are equivalent from an ordering perspective.
 (this didn't always work in the past though, because we had a bug in that logic)